src/bd/primitives/PrimitiveLoader.java

```diff
@@ -0,0 +1,113 @@
+   * the primitive allows in-parser specialization, and the argument nodes match the
+   * expectations, than a specializer is returned. otherwise, null is returned.
+   */
+  @SuppressWarnings("unchecked")
+  public final Specializer<EagerT, Context, ExprT> getParserSpecializer(final Id selector,
+      final ExprT[] argNodes) {
+    Specializer<? extends ExprT, Context, ExprT> specializer = eagerPrimitives.get(selector);
+    if (specializer != null && specializer.inParser() && specializer.matches(null, argNodes)) {
+      return (Specializer<EagerT, Context, ExprT>) specializer;
+    }
+    return null;
+  }
+
+  /**
+   * Lookup a specializer for specialization during execution.
+   *
+   * <p>
+   * If one is found for the given selector/id, it is checked whether the run-time arguments as
+   * well as the argument nodes match for the specialization. If they match, the specializer is
+   * returned, null is returned otherwise.
+   */
+  @SuppressWarnings("unchecked")
+  public final Specializer<EagerT, Context, ExprT> getEagerSpecializer(final Id selector,
+      final Object[] arguments, final ExprT[] argumentNodes) {
+    Specializer<? extends ExprT, Context, ExprT> specializer =
+        eagerPrimitives.get(selector);
+    if (specializer != null && specializer.matches(arguments, argumentNodes)) {
+      return (Specializer<EagerT, Context, ExprT>) specializer;
+    }
+    return null;
+  }
+
+  /**
+   * Create a {@link Specializer} for the given {@link Primitive}.
+   */
+  @SuppressWarnings("unchecked")
+  protected final <T> Specializer<T, Context, ExprT> createSpecializer(final Primitive prim,
+      final NodeFactory<? extends ExprT> factory) {
+    try {
+      // Try with erased type signature
+      return prim.specializer()
+                 .getConstructor(Primitive.class, NodeFactory.class, Object.class)
+                 .newInstance(prim, factory, context);
+    } catch (InstantiationException | IllegalAccessException | IllegalArgumentException
+        | InvocationTargetException | SecurityException e) {
+      throw new RuntimeException(e);
+    } catch (NoSuchMethodException e) {
+      try {
+        // Try with concrete type signature
+        return prim.specializer()
+                   .getConstructor(Primitive.class, NodeFactory.class, context.getClass())
+                   .newInstance(prim, factory, context);
+      } catch (InstantiationException | IllegalAccessException | IllegalArgumentException
+          | InvocationTargetException | NoSuchMethodException | SecurityException e1) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  /**
+   * Get the {@link Primitive} annotation from a {@link NodeFactory}.
+   */
+  protected final Primitive[] getPrimitiveAnnotation(
+      final NodeFactory<? extends ExprT> factory) {
+    Class<?> nodeClass = factory.getNodeClass();
+    return nodeClass.getAnnotationsByType(Primitive.class);
+  }
```

Comment:
Why not? It is common code used by the subclasses?
Where else would you put it?